sno/init.py

```diff
@@ -503,6 +520,82 @@ def build_meta_info(self):
         yield from gpkg.get_meta_info(db, layer=self.table)
 
 
+class ImportPostgreSQL(OgrImporter):
+    @classmethod
+    def postgres_url_to_ogr_conn_str(cls, url):
+        """
+        Takes a URL ('postgresql://..')
+        and turns it into a key/value connection string, prefixed by 'PG:' for OGR.
+
+        libpq actually handles URIs fine, but OGR doesn't :(
+        So to import via OGR we have to convert them.
+        https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING
+
+        We don't currently handle all of the things these URLs can contain.
+        In particular, we don't (currently) handle:
+         * domain sockets
+         * multiple hostnames or ports
+        """
+
+        url = urlsplit(url)
+        scheme = url.scheme.lower()
+        if scheme not in ('postgres', 'postgresql'):
+            raise ValueError("Bad scheme")
+
+        params = parse_qs(url.query)
+
+        if url.username:
+            params.setdefault('user', url.username)
+        if url.password:
+            params.setdefault('password', url.password)
+        if url.hostname:
+            params.setdefault('host', url.hostname)
+        if url.port:
+            params.setdefault('port', url.port)
+        dbname = (url.path or '/')[1:]
+        if dbname:
+            params.setdefault('dbname', dbname)
+
+        conn_str = ' '.join(sorted(f'{k}={v}' for (k, v) in params.items()))
+        return f'PG:{conn_str}'
+
+    @classmethod
+    def handle_source_string(cls, source):
+        if '://' not in source:
+            return None
+        try:
+            return cls.postgres_url_to_ogr_conn_str(source), ['PG']
+        except ValueError:
+            return None
+
+    def psycopg2_conn(self):
+        conn_str = self.source
+        if conn_str.startswith('OGR:'):
+            conn_str = conn_str[4:]
+        if conn_str.startswith('PG:'):
+            conn_str = conn_str[3:]
+        # this will either be a URL or a key=value conn str
+        return psycopg2.connect(conn_str)
+
+    @property
+    @functools.lru_cache(maxsize=1)
+    def primary_key(self):
+        conn = self.psycopg2_conn()
+        with conn.cursor() as cur:
+            cur.execute(
+                """
+                SELECT a.attname
+                FROM   pg_index i
+                JOIN   pg_attribute a ON a.attrelid = i.indrelid
+                                     AND a.attnum = ANY(i.indkey)
+                WHERE  i.indrelid = %s::regclass
+                AND    i.indisprimary;
+                """,
+                [self.table],
+            )
+            return cur.fetchone()[0]
```

Comment:
if there's a multi-column PK? At a minimum should error out, not silently use one of them. That'd be bad.